core/src/main/java/org/bitcoinj/wallet/FriendKeyChain.java

```diff
@@ -0,0 +1,271 @@
+    public static final int PATH_INDEX_FROM_ID = 6;
+
+
+    public FriendKeyChain(DeterministicSeed seed, ImmutableList<ChildNumber> rootPath, int account, Sha256Hash myBlockchainUserId, Sha256Hash theirBlockchainUserId) {
+        super(seed, ImmutableList.<ChildNumber>builder().addAll(rootPath).add(new ChildNumber(account, true)).add(new ExtendedChildNumber(myBlockchainUserId)).add(new ExtendedChildNumber(theirBlockchainUserId)).build());
+        type = KeyChainType.RECEIVING_CHAIN;
+    }
+
+    public FriendKeyChain(DeterministicSeed seed, KeyCrypter keyCrypter, ImmutableList<ChildNumber> rootPath,  int account, Sha256Hash myBlockchainUserId, Sha256Hash theirBlockchainUserId) {
+        super(seed, keyCrypter, ImmutableList.<ChildNumber>builder().addAll(rootPath).add(new ChildNumber(account, true)).add(new ExtendedChildNumber(myBlockchainUserId)).add(new ExtendedChildNumber(theirBlockchainUserId)).build());
+        type = KeyChainType.RECEIVING_CHAIN;
+    }
+
+    public FriendKeyChain(DeterministicSeed seed, ImmutableList<ChildNumber> path) {
+        super(seed, path);
+        type = KeyChainType.RECEIVING_CHAIN;
+    }
+
+    public FriendKeyChain(DeterministicSeed seed, KeyCrypter keyCrypter, ImmutableList<ChildNumber> path) {
+        super(seed, keyCrypter, path);
+        type = KeyChainType.RECEIVING_CHAIN;
+    }
+
+    /** xpub must contain a depth to indicate that it is not a masterkey **/
+    public FriendKeyChain(NetworkParameters params, String xpub, EvolutionContact contact) {
+        super(DeterministicKey.deserializeB58(xpub, getContactPath(params, contact, KeyChainType.SENDING_CHAIN), params));
+        setAccountPath(getContactPath(params, contact, KeyChainType.SENDING_CHAIN));
+        //TODO:  check to see that xpub has a depth, if not the path won't agree with getContactPath
+    }
+
+    public FriendKeyChain(DeterministicKey watchingKey) {
+        super(watchingKey);
+    }
+
+    public FriendKeyChain(DeterministicKey watchingKey, boolean isFollowing) {
+        super(watchingKey, isFollowing);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public DeterministicKey getKey(KeyPurpose purpose) {
+        return getKeys(purpose, 1).get(0);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public List<DeterministicKey> getKeys(KeyPurpose purpose, int numberOfKeys) {
+        checkArgument(numberOfKeys > 0);
+        lock.lock();
+        try {
+            DeterministicKey parentKey;
+            int index;
+            switch (purpose) {
+                case RECEIVE_FUNDS:
+                case REFUND:
+                    issuedKeys += numberOfKeys;
+                    index = issuedKeys;
+                    parentKey = getKeyByPath(getAccountPath());
+                    break;
+                default:
+                    throw new UnsupportedOperationException();
+            }
+            // Optimization: potentially do a very quick key generation for just the number of keys we need if we
```

Comment:
The same comment was removed from AuthenticationKeyChain, do we need it here?

Reply:
Not sure, need to review.  The comment is also in DeterministicKeyChain.getKeys

Reply:
https://github.com/dashevo/dashj/pull/49/commits/a2b9daa4777312b1f11d05cc2a984a993179ebe6 comments make reference to the DeterministicKeyChain.getKeys method.